lib/redirect-pool.js

```diff
@@ -0,0 +1,149 @@
+      return headers[i + 1]
+    }
+  }
+
+  return null
+}
+
+// https://tools.ietf.org/html/rfc7231#section-6.4.4
+function shouldRemoveHeader (header, removeContent) {
+  const lcHeader = header.toLowerCase()
+
+  return lcHeader === 'host' || (removeContent && lcHeader.indexOf('content-') === 0)
+}
+
+// https://tools.ietf.org/html/rfc7231#section-6.4
+function cleanRequestHeaders (headers, removeContent) {
+  if (Array.isArray(headers)) {
+    for (let i = headers.length - 2; i >= 0; i -= 2) {
+      if (shouldRemoveHeader(headers[i], removeContent)) {
+        headers.splice(i, 2)
+      }
+    }
+  } else {
+    // IncomingHttpHeaders
+    for (const header of Object.keys(headers)) {
+      if (shouldRemoveHeader(header, removeContent)) {
+        headers[header] = undefined
+      }
+    }
+  }
+}
+
+class RedirectPool extends Pool {
+  constructor (origin, options) {
+    super(origin, options)
+
+    this[kOrigin] = origin
+  }
+
+  dispatch (opts, handler) {
+    const pool = this
+
+    // Compute the number of left redirects, if needed
+    if (!(kLeftRedirections in opts)) {
+      opts[kLeftRedirections] = parseMaxRedirections(opts)
+    }
+
+    const originalOnHeaders = handler.onHeaders
+
+    handler.onHeaders = function onHeadersWithRedirect (statusCode, headers, resume) {
+      // Check if statusCode is 3xx, if there is a location header and if the redirection can be followed
+      const location = redirectLocation(statusCode, headers, opts)
+
+      // Nothing to follow, use the original implementation
+      if (!location) {
+        return originalOnHeaders.call(handler, statusCode, headers, resume, { redirections: opts[kFollowedRedirections] })
+      }
+
+      /*
+        https://tools.ietf.org/html/rfc7231#section-6.4
+
+        Remove headers referring to the original URL.
+        By default it is Host only, unless it's a 303 (see below), which removes also all Content-* headers.
+      */
+      if (opts.headers) {
+        cleanRequestHeaders(opts.headers, statusCode === 303)
+      }
+
+      // https://tools.ietf.org/html/rfc7231#section-6.4.4
+      // In case of HTTP 303, always replace method to be either HEAD or GET
+      if (statusCode === 303 && opts.method !== 'HEAD') {
+        opts.method = 'GET'
+      }
+
+      // Add the current URL to the list of redirects
+      if (!opts[kFollowedRedirections]) {
+        opts[kFollowedRedirections] = []
+      }
+
+      opts[kFollowedRedirections].push(`${pool[kOrigin]}${opts.path}`)
+
+      // Update options
+      opts.path = null
+      opts[kLeftRedirections]--
+
+      // Follow the redirect
+      switch (opts.requestType) {
+        case 'request':
+          request(location, opts, handler.callback)
+          break
+        case 'stream':
+          stream(location, opts, handler.factory, handler.callback)
+          break
+        case 'pipeline':
+          pipeline(location, opts, handler.handler)
+          break
+      }
```

Comment:
I don’t think this should be needed.

Reply:
I need this in order to support cross origin redirection.

Reply:
Removed the part about `pipeline` (since RedirectPool is not supported on pipeline).
About `request` vs `stream`, I know try to autodetect given the shape of the handler.